### ISSUE TITLE
Docker image for swig

### ISSUE THREAD

**Issue by vaibhavsood** [swig/swig#1112]
Hi,

Want to check if there is a docker image/Dockerfile for swig which is officially supported? I could not find one under dockerhub official images https://hub.docker.com/explore/

If not, want to check if there is any plan to add an official swig image to dockerhub? https://docs.docker.com/docker-hub/official_repos/#how-do-i-create-a-new-official-repository

**Comment by ojwb** [comment]
AFAIK, there isn't and there are no such plans.

What do you see it being useful for?  SWIG isn't some complex platform or web application - it's a single self-contained tool, and a whole docker image for a single tool doesn't seem a helpful approach.

We also already have more to do than there is developer time for.

**Comment by wsfulton** [comment]
I also don't know of any plans. I'm quite happy to support an official image though if @vaibhavsood or someone can maintain it and ensure the quality is 1st class. The creation and testing of the image would need to be automated in some way. I have an experimental dockerfile I once created for a docker image with gcc-7 to test SWIG, but you probably know more what should be in the dockerfile. 

- Would it make sense to test the image on Travis? (Travis does support Docker images.) 
- What would go on the image, given that SWIG is rather useless without a C/C++ compiler and target languages? SWIG can work with about 20 target languages atm. Would all the languages and compilers be installed or would it be just the tools required to build SWIG?

**Comment by ojwb** [comment]
I guess a "SWIG development" image would have its uses.  However, trying to keep an image with ~20 different languages up to date seems like a lot of on-going work, so it really needs a volunteer who is actually going to commit to doing all that work for a useful length of time, or else we'll quickly end up with an image full of out of date versions of things and a growing pile of support queries about it.

**Comment by rcoup** [comment]
FWIW, I've been using this container image for running specific versions of SWIG. Helps to avoid massive diffs when projects aren't using the same version as you have installed.

Take/use/adapt/whatever 👍 

https://gist.github.com/rcoup/35e11e31d069689c9c09a70cc2fc7a0d

**Comment by wsfulton** [comment]
Building comments made here, there are quite a few different ways to develop on Windows, from different compilers, different IDEs such as Visual Studio code vs Visual Studio 20xx, different versions of compilers (gcc, cccl, cl, ...) and unix-like layers cygwin, msys etc. This is a constantly changing landscape in Windows too and developers have their own personal preferences. There are also so many different SWIG target languages and some are not working on Windows. There is limited test-suite support in Windows and Linux is where serious development of SWIG happens. For these reasons I suggest the Docker images should be community maintained. #2161 has a similar Docker proposal. I suggest the wiki is used to point to Docker images that meet various needs. Add a link from https://github.com/swig/swig/wiki/Getting-Started and please add full instructions.

If someone puts together a comprehensive CI test of a Docker image on say Github Actions or CircleCI in a pull request, then perhaps we can adopt it as an official image. The pull requester will need to offer to maintain it going forwards.

**Comment by romintomasetti** [comment]
Hi @wsfulton !

Well, looking at you CICD pipeline, you're already using quite some different compilers/version of them to compile SWIG itself. 

Maintaining dockerfiles for each compiler plus for each supported target language would be a nightmare for sure.

So, why not building one single image, using the latest Debian for instance, with whatever default gcc comes with it, and install also the default Python, Python 3, etc.

You would then add a single job in your CICD that runs in this image (both compiling SWIG and launching the test suite).

This image/job would be useful for:
* developers that want to work in a containerized environment
* show that at least you support Debian
* ensure that the list of requirements is up-to-date (both for compiling SWIG or using supported target languages)

What do you think ?

**Comment by wsfulton** [comment]
Our CI/CD pipeline is extensive and takes quite a bit of maintenance already and a lot of time to set it up. It already covers your last two points. While some of our CI could be replaced by Docker, it'll be quite an undertaking to cover all we already have. We make extensive use of Gitlab compute resources and I don't think it makes sense have repeated testing. Running one long test-suite is not as good as parallelising it the way we currently do as it'll hit a bunch of time limits. The instructions to get going are mostly a bunch of `apt-get install` commands and these are documented at https://github.com/swig/swig/wiki/Getting-Started and any competent developer will be up and running within a few minutes.

So in summary, a Docker image would be more a nice-to-have, but the current set of SWIG developers are truly stretched to consider this unless someone offers to help and free us of the additional maintenance burden. BTW, a lot is done outside of the SWIG project, for example, Redhat are continually testing updated dependencies in Fedora and report any problems they see well before regular users get to see them.

**Comment by ojwb** [comment]
> For these reasons I suggest the Docker images should be community maintained.

There are also other container-like systems as well as Docker, and the SWIG development team is already overstretched, so I agree it makes more sense for such images to be community maintained - closing.
